src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetCommandHandler.cs

```diff
@@ -51,6 +51,12 @@ protected virtual bool TryInvokeSnippetPickerOnQuestionMark(ITextView textView,
             return false;
         }
 
+        protected virtual Func<char, bool> GetIsIdentifierPartCharacterFunction(Document document)
```

Comment:
Why have a virtual function returning a function? Why can't this just be IsIdentiferPartCharacter(char)?

(I'd also ask even bigger questions like "should this be calling text structure navigator" but I digress.)

Reply:
The issue is that `ISyntaxFactsService` is obtained from the document which is derived from the arguments to `TryHandleTypedSnippet()`.  This _could_ be implemented as `bool IsIdentifierTypedCharacter(char c, Document d)`, but that would require a call to `.GetService<>()` for every character.

Reply:
is the issue that ISyntaxFactService has dependency to our SyntaxNode? can we split ISyntaxFactSerivce from one that has dependency to SyntaxNode to ones that doesn't? and you guys implement the part that doesnt?

Reply:
the primary problem is that this piece of logic in the snippet command handler is language specific.  it's effectively trying to find the start of the language construct, and using that to determine which snippet name to select.

Reply:
Note: we could just make this completely agnostic and just say:

we walk backwards, consuming every letter, number, connecting char, combining char, formatting char, and # and ~.  And just not defer to the language at all here.